vendor/parse-wyc/lib/parse_wyc.ml

```diff
@@ -0,0 +1,238 @@
+            Intermediate (Recovering (R.generate env)) )
+    | Recovering candidates -> (
+        match R.attempt candidates token with
+        | `Ok (cp, _) -> Intermediate (Correct (M.recover cp))
+        | `Accept x -> Success x
+        | `Fail -> (
+            match token with
+            | Parser.EOF, _, _ -> (
+                match candidates.final with
+                | None -> Error
+                | Some x -> Success x )
+            | _ -> Intermediate parser ) )
+end
+
+let parse_with_recovery entrypoint tokens =
+  let module P = With_recovery in
+  let rec step tokens = function
+    | P.Error -> failwith "Parsing failed"
+    | P.Success x -> x
+    | P.Intermediate p -> offer p tokens
+  and offer p tokens =
+    let token, rest =
+      match tokens with
+      | [] -> ((Parser.EOF, Lexing.dummy_pos, Lexing.dummy_pos), [])
+      | token :: rest -> (token, rest)
+    in
+    step rest (P.step p token)
+  in
+  offer (P.initial entrypoint Lexing.dummy_pos) tokens
+
+let lex_buf lexbuf =
+  Lexer.init ();
+  let rec loop acc =
+    match Lexer.token lexbuf with
+    | exception Lexer.Error _ -> loop acc
+    | token -> (
+        let acc = (token, lexbuf.lex_start_p, lexbuf.lex_curr_p) :: acc in
+        match token with Parser.EOF -> List.rev acc | _ -> loop acc )
+  in
+  loop []
+
+let merge_adj merge l =
+  List.fold_left
+    (fun acc x ->
+      match acc with
+      | h :: t -> (
+          match merge h x with Some m -> m :: t | None -> x :: h :: t )
+      | [] -> x :: acc)
+    [] l
+  |> List.rev
+
+let normalize_locs locs =
+  List.sort_uniq Location.compare locs |> merge_adj Location.merge
+
+let invalid_locs p m lexbuf =
+  let ast = parse_with_recovery p (lex_buf lexbuf) in
+  let loc_stack = Stack.create () in
+  let loc_list = ref [] in
+  let make_mapper () =
+    let open Parsetree in
+    let default = Ast_mapper.default_mapper in
+    let expr m e =
+      if Annot.Exp.is_generated e then
+        loc_list := Stack.top loc_stack :: !loc_list;
+      default.expr m e
+    in
+    let class_expr m x =
+      if Annot.Class_exp.is_generated x then
+        loc_list := Stack.top loc_stack :: !loc_list;
+      default.class_expr m x
+    in
+    let wrap mapper loc f x =
+      if Stack.is_empty loc_stack then (
+        Stack.push loc loc_stack;
+        let x = f mapper x in
+        ignore (Stack.pop loc_stack);
+        x )
+      else f mapper x
+    in
+    let structure_item m x = wrap m x.pstr_loc default.structure_item x in
+    let signature_item m x = wrap m x.psig_loc default.signature_item x in
+    let attribute m x =
+      if Annot.Attr.is_generated x then
+        loc_list := Stack.top loc_stack :: !loc_list;
+      default.attribute m x
+    in
+    {
+      Ast_mapper.default_mapper with
+      attribute;
+      expr;
+      class_expr;
+      structure_item;
+      signature_item;
+    }
+  in
+  let mapper = make_mapper () in
+  ignore ((m mapper) ast);
+  normalize_locs !loc_list
```

Comment:
This function is collecting the locs of invalid parts of the code.
But we have an ast here, why not use it instead of going back to a string in the next function ?
The `Ast_mapper` and `Stack` tricks used here can be used to map "structure item that contains an annot, deeply" to "invalid ast ext node" directly.

Reply:
I thought to have a string output was a simpler interface for ocamlformat, as it is what is processed in Translation_unit, and returning the ast types is tricky as we have an OMP layer which is not the same as the one of ocamlformat (Migrate_ast is duplicated)

Reply:
Having two different AST/parser seems like something that could cause bugs. How hard to implement and what are the drawback to use a single implementation ?

Having control on the parser's code may solve a few issues too (eg. to modify it or to avoid changes in behavior between OMP's versions, which could cause formatting changes between the same version of OCamlformat)

Reply:
It's necessary if we want this parser (pwyc) to be independent from ocamlformat, which is the what we agreed upon. Maybe the Migrate_ast could be factorized in the future, but I don't think it should be done in this PR.

Reply:
My question was about using the AST and parser from pwyc in OCamlformat, dropping OMP. The dependency stays the same. 
My question is, what are the drawbacks of doing that ?